testing/sdk_auth.py

```diff
@@ -0,0 +1,376 @@
+        in the following order:
+        - primary: first part of the principal. In the case of a user, it's the same as your username.
+                   For a host, the primary is the word host.
+        - instance: The instance is a string that qualifies the primary.
+                    In the case of a user, the instance is usually null, but a user might also have an additional
+                    principal, with an instance called admin.  In the case of a host, the instance is the fully
+                    qualified hostname, e.g., daffodil.mit.edu.
+        - realm: your Kerberos realm. In most cases, your Kerberos realm is your domain name, in upper-case letters.
+
+        More info on principal definition:
+        https://web.mit.edu/kerberos/krb5-1.5/krb5-1.5.4/doc/krb5-user/What-is-a-Kerberos-Principal_003f.html
+
+        A principal is formatted as: <primary>/instance@realm
+        Eg. hdfs/name-0-node.hdfs.autoip.dcos.thisdcos.directory@LOCAL
+
+        :param principals: The list of principals to be added to KDC.
+        """
+        # TODO: Perform sanitation check against validity of format for all given principals and raise an
+        # exception when the format of a principal is invalid.
+        self.principals = principals
+
+        log.info("Adding the following list of principals to the KDC: {principals}".format(principals=principals))
+        add_principal_cmd = "task exec {task_id} /usr/sbin/kadmin -l add --use-defaults --random-password {principal}"
+        for principal in principals:
+            try:
+                sdk_cmd.run_cli(add_principal_cmd.format(task_id=KERBEROS_APP_ID, principal=principal))
+            except dcos.errors.DCOSException as e:
+                raise RuntimeError("Failed to add principal {principal}: {err_msg}".format(
+                    principal=principal, err_msg=repr(e)))
+
+        log.info("Principals successfully added to KDC")
+
+    def __create_and_fetch_keytab(self):
+        """
+        Creates the keytab file that holds the info about all the principals that have been
+        added to the KDC. It also fetches it locally so that later the keytab can be uploaded to the secret store.
+        """
+        log.info("Creating the keytab")
+        cmd = "task exec {task_id} /usr/sbin/kadmin -l ext -k {keytab_name} ".format(
+            task_id=KERBEROS_APP_ID, keytab_name=self.keytab_file_name
+        )
+        cmd = cmd + ' '.join(principal for principal in self.principals)
+        try:
+            sdk_cmd.run_cli(cmd)
+        except Exception as e:
+            raise(e)
+
+        _copy_file_to_localhost(self)
+
+    def __create_and_upload_secret(self):
+        """
+        This method base64 encodes the keytab file and creates a secret with this encoded content so the
+        tasks can fetch it.
+        """
+        log.info("Creating and uploading the keytab file to the secret store")
+
+        cmd = "package install --yes --cli dcos-enterprise-cli"
```

Comment:
Question: is there a benefit in doing this here, instead of as part of the test setup or earlier in the process? Here we may install the enterprise CLI multiple times.

Reply:
No benefit, just did it as part of the context of dealing with secrets.

Moved to end of constructor.